contracts/Certificate.sol

```diff
@@ -61,6 +61,26 @@ contract Certificate is
 {
   using UInt256ArrayLib for uint256[];
 
+  /**
+   * @notice The data that is passed to the `onERC1155BatchReceived` function data parameter when creating a new
+   * certificate.
+   * @dev This struct is used to pass data to the `onERC1155BatchReceived` function when creating a new certificate.
+   *
+   * @param isReplacement A bool used to differentiate between a token batch being received to create a new
+   * certificate and a token batch being received as a replacement for previously released removals.
+   * @param recipient The address is the address that will receive the new certificate.
+   * @param certificateAmount The amount of the certificate that will be minted.
+   * @param purchasingTokenAddress The address is the address of the token that was used to purchase the certificate.
+   * @param priceMultiple The number of purchasing tokens required to purchase one NRT.
+   */
+  struct CertificateData {
+    bool isReplacement;
+    address recipient;
+    uint256 certificateAmount;
+    address purchasingTokenAddress;
+    uint256 priceMultiple;
+  }
```

Comment:
can probably use a smaller uint here which would allow for tighter packing in the struct. Currently at 840 bits I believe (3.2x 256 bit words). 

```suggestion
  struct CertificateData {
    uint256 certificateAmount; // 256
    address purchasingTokenAddress; // 160
    bool isReplacement; // 8
    uint88 priceMultiple;  // 88 (this + prev 2 == 256)
    address recipient; // 160
  }
```

That should make for 3 256 bit chunks instead of 4

Reply:
great - implemented this but reordered things to keep the bool as the first byte because we need to decode that for determining what kind of batch received this -- new cert or replacement removals

Reply:
actually never mind, TWL that struct packing only leads to gas benefits when the structs are being written to or read from storage